### PR TITLE
OpInfo porting for torch.real & torch.imag

### PR DESCRIPTION
**Pull request by Kiyosora**
Related #54298

This PR ports the method_tests() entries of torch.real & torch.imag to OpInfo.

### CODE REVIEW

**Comment by mruberry** [comment]
Hey @Kiyosora! Thanks for this PR. I've requested a review from @anjali411, who'll be able to help you out.

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -3228,6 +3232,26 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                        SkipInfo('TestUnaryUfuncs', 'test_reference_numerics_hard',
                                 device_type='cpu', dtypes=[torch.complex64])
                    )),
+    OpInfo('real',
+           op=torch.real,
+           dtypes=complex_types(),
+           dtypesIfCPU=complex_types(),
```
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:3238]:
```suggestion
           
```

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -3228,6 +3232,26 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
+           dtypesIfCUDA=complex_types(),
```
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:3239]:
```suggestion

```

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -3228,6 +3232,26 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
+           skips=(
+               SkipInfo('TestCommon', 'test_variant_consistency_eager'),
```
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:3241]: Hey @Kiyosora why do we need these skips? could you share the error message for the failing tests?

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -3228,6 +3232,26 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
+               SkipInfo('TestCommon', 'test_variant_consistency_jit'),
+           ),
+           sample_inputs_func=sample_inputs_real_imag,),
+    OpInfo('imag',
+           op=torch.imag,
+           dtypes=complex_types(),
+           dtypesIfCPU=complex_types(),
+           dtypesIfCUDA=complex_types(),
```
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:3249]:
```suggestion
```

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:3239]:
```suggestion
```

**Comment by anjali411** [comment]
Just realized, there's another PR similar to this one https://github.com/pytorch/pytorch/pull/51946, so I am gonna let you and @kshitij12345 decide which one should be merged.

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:3239]: I think it is because `test_variant_consistency_eager` will fail because real and imag are attributes.

Relevant conversation from the previous PR.

https://github.com/pytorch/pytorch/pull/51946#discussion_r573546348

**Comment by kshitij12345** [comment]
@anjali411 @Kiyosora will close the older PR in favour of this :)

**Comment by Kiyosora** [comment]
@kshitij12345 @anjali411 Thank you very much! Just as you analysed, The 2 skipped cases would caused the error of `TypeError: 'getset_descriptor' object is not callable` .
I'm so sorry that I have not notice that there is already has a same PR before, I would absorb the changes from the old PR and complete it here.

File: test/test_ops.py [hunk 5]
```diff
@@ -271,6 +271,10 @@ def test_variant_consistency_eager(self, device, dtype, op):
                 if variant in inplace_ops and skip_inplace:
                     continue
 
+                if not callable(variant):
```
Comment (kshitij12345) [inline test/test_ops.py:274]: IIRC as per the discussion in previous PR, plan was to update following lines if the operator is not callable.
https://github.com/pytorch/pytorch/blob/1b4bb3691cbddfea45805b8459292842e5b23f17/torch/testing/_internal/common_methods_invocations.py#L191-L195

This way we won't have to update anything in `test_ops` as they would get filtered at following lines automatically.
https://github.com/pytorch/pytorch/blob/1b4bb3691cbddfea45805b8459292842e5b23f17/test/test_ops.py#L238

https://github.com/pytorch/pytorch/blob/1b4bb3691cbddfea45805b8459292842e5b23f17/test/test_ops.py#L308-L309


I think you might want to discuss and finalise the approach once.

@anjali411 @mruberry

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -3393,6 +3397,18 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                        SkipInfo('TestUnaryUfuncs', 'test_reference_numerics_hard',
                                 device_type='cpu', dtypes=[torch.complex64])
                    )),
+    OpInfo('real',
```
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:3400]: I'd suggest using `UnaryUfuncInfo` with corresponding numpy functions for reference.

That way we will get tests verifying the output of these function against numpy equivalents and multiple other tests (look at test_unary_ufuncs.py for reference)

You might have to tweak one or two functions there (you can refer the previous [PR](https://github.com/pytorch/pytorch/pull/51946/files) for some reference to the changes).

File: test/test_ops.py (hunk 5 above)
Comment (anjali411) [reply test/test_ops.py:274]: yeah this sounds good!

Comment (anjali411) [reply test/test_ops.py:274]: yup as @kshitij12345 mentioned, we should set the method_variant to `None` if the operator is not callable!

**Comment by anjali411** [comment]
@Kiyosora could you please rebase this PR?

**Review by anjali411 — APPROVED** [review]
LGTM thanks @Kiyosora ! could you please also remove real and imag from this list? https://github.com/pytorch/pytorch/blob/5f90ed550ccffbe02476a9229ecddc3dc89b3d50/test/test_autograd.py#L5263

File: test/test_unary_ufuncs.py [hunk 7]
```diff
@@ -479,7 +479,7 @@ def _test_out_arg(self, op, input, output, expected, **kwargs):
             self.assertTrue(res is output)
             self.assertEqual(output, expected.to(output.dtype))
 
-    @ops(unary_ufuncs, dtypes=OpDTypes.supported)
+    @ops(list(filter(lambda op: op.supports_out, unary_ufuncs)), dtypes=OpDTypes.supported)
```
Comment (anjali411) [inline test/test_unary_ufuncs.py:482]: @Kiyosora is this change intended?

Comment (kshitij12345) [reply test/test_unary_ufuncs.py:482]: https://github.com/pytorch/pytorch/blob/b5647dd52b48a51fec1387382deb5b59c7651512/aten/src/ATen/native/native_functions.yaml#L280-L284

`real` and `imag` don't have out variants. However the `UnaryUfuncInfo` assumes that every unary operator has `out` (which is true except for these two). 
So for this `test_out_arg_all_dtypes` which tests `out` variant, we have to filter these two operators.

**Note**: `list` is important. `filter` returns generator which is emptied with the tests for first device_type. cc: @mruberry

Comment (mruberry) [reply test/test_unary_ufuncs.py:482]: Raising a skip if the operation does not support out in the test with a message explaining the skip is probably simpler than implementing the filter in the decorator.

Comment (kshitij12345) [reply test/test_unary_ufuncs.py:482]: Right. That makes sense!

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -2979,6 +2979,19 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
+                   dtypes=complex_types(),
+                   dtypesIfCPU=complex_types(),
+                   dtypesIfCUDA=complex_types(),
+                   dtypesIfROCM=complex_types(),
+                   supports_out=False,
+                   supports_autograd=False,
+                   skips=(
+                       # Skip since real and imag don't have out variants.
+                       SkipInfo('TestUnaryUfuncs', 'test_out_arg_all_dtypes',
+                                dtypes=[torch.cfloat, torch.cdouble]),
```
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:2993]: minor: you can skip adding `dtypes` as this applicable to the operator as a whole.

Comment (Kiyosora) [reply torch/testing/_internal/common_methods_invocations.py:2993]: Thanks for reminding! Let me fix it ; )

**Review by kshitij12345 — APPROVED** [review]
LGTM!

**Comment by anjali411** [comment]
@Kiyosora could you please rebase this PR?
